apps/pipelines/nodes/nodes.py

```diff
@@ -120,6 +129,33 @@ def _get_context(self, input, state: PipelineState, prompt: PromptTemplate):
 
         return context
 
+    def _get_history(self, session: ExperimentSession, node_id: str):
+        if self.history_type == PipelineChatHistoryTypes.NAMED:
+            history_name = self.history_name
+        elif self.history_type == PipelineChatHistoryTypes.GLOBAL:
+            history_name = "global"
+        else:
+            history_name = node_id
+
+        try:
+            history: PipelineChatHistory = session.pipeline_chat_history.get(type=self.history_type, name=history_name)
```

Comment:
could this be pushed into the state class so that it can be cached?

Reply:
Do you mean into the `PipelineState` or some other state class? 

Yes, we can totally do that - however, it will only benefit when you have more two or more nodes with the same history name in the pipeline (since the pipeline is invoked with a new `PipelineState` for every message).

This will probably best be done after the refactor that @SmittieC was suggesting (moving the `PipelineState` into the invoke method), as you can't add arbitrary items onto a langchain state class and have them persist through the pipeline without them also being added to the first call. i.e. when invoking the pipeline we'd need to call it like this everywhere:

```
runnable.invoke(PipelineState(cached_history={}, messages=[user_input], experiment_session=experiment_session))
```

If it's OK with you, I might punt this one down the line as it will make the diff quite a bit larger on this PR. 

Here is what it would look like: https://github.com/dimagi/open-chat-studio/commit/9d3197889a5c791c8f0c5a2b5cb9db7917db7238

Reply:
Fine to punt. On further consideration I'm not actually sure it's even relevant for our current use cases since even in the case of shared history, only one node would execute at a time. It would only make a difference if there were multiple nodes places one after another which used the same history.

Reply:
Yes, it's only valid if you have two or more nodes with the same history in the same pipeline.